pkg/kernel/kernel.go

```diff
@@ -0,0 +1,311 @@
+	if err := receipt.PublishKernelStatus(kernelBusy); err != nil {
+		log.Printf("Error publishing kernel status 'busy': %v\n", err)
+	}
+	defer func() {
+		if err := receipt.PublishKernelStatus(kernelIdle); err != nil {
+			log.Printf("Error publishing kernel status 'idle': %v\n", err)
+		}
+	}()
+
+	switch receipt.Msg.Header.MsgType {
+	case "kernel_info_request":
+		if err := k.sendKernelInfo(receipt); err != nil {
+			log.Fatal(err)
+		}
+	case "complete_request":
+		if err := k.handleCompleteRequest(receipt); err != nil {
+			log.Fatal(err)
+		}
+	case "execute_request":
+		if err := k.handleExecuteRequest(receipt); err != nil {
+			log.Fatal(err)
+		}
+	case "shutdown_request":
+		k.handleShutdownRequest(receipt)
+	default:
+		log.Println("Unhandled shell message: ", receipt.Msg.Header.MsgType)
+	}
+}
+
+// sendKernelInfo sends a kernel_info_reply message.
+func (k Kernel) sendKernelInfo(receipt msgReceipt) error {
+	return receipt.Reply("kernel_info_reply",
+		Info{
+			ProtocolVersion:       k.ProtocolVersion,
+			Implementation:        "Kubeflow",
+			ImplementationVersion: k.Version,
+			Banner:                fmt.Sprintf("Kubeflow kernel: Kubeflow - v%s", k.Version),
+			LanguageInfo: LanguageInfo{
+				Name:          "Python (Kubeflow)",
+				Version:       runtime.Version(),
+				FileExtension: ".py",
+			},
+			HelpLinks: []HelpLink{
+				{Text: "Kubeflow", URL: "https://www.kubeflow.org/"},
+			},
+		},
+	)
+}
+
+// handleExecuteRequest runs code from an execute_request method,
+// and sends the various reply messages.
+func (k Kernel) handleExecuteRequest(receipt msgReceipt) error {
+
+	// Extract the data from the request.
+	reqcontent := receipt.Msg.Content.(map[string]interface{})
+	code := reqcontent["code"].(string)
+	silent := reqcontent["silent"].(bool)
+
+	if !silent {
+		k.ExecCounter++
+	}
+
+	// Prepare the map that will hold the reply content.
+	content := make(map[string]interface{})
+	content["execution_count"] = k.ExecCounter
+
+	// Tell the front-end what the kernel is about to execute.
+	if err := receipt.PublishExecutionInput(k.ExecCounter, code); err != nil {
+		log.Printf("Error publishing execution input: %v\n", err)
+	}
+
+	// Redirect the standard out from the REPL.
+	oldStdout := os.Stdout
+	rOut, wOut, err := os.Pipe()
+	if err != nil {
+		return err
+	}
+	os.Stdout = wOut
+
+	// Redirect the standard error from the REPL.
+	oldStderr := os.Stderr
+	rErr, wErr, err := os.Pipe()
+	if err != nil {
+		return err
+	}
+	os.Stderr = wErr
+
+	var writersWG sync.WaitGroup
+	writersWG.Add(2)
+
+	// Forward all data written to stdout/stderr to the front-end.
+	go func() {
+		defer writersWG.Done()
+		jupyterStdOut := JupyterStreamWriter{StreamStdout, &receipt}
+		io.Copy(&jupyterStdOut, rOut)
+	}()
+
+	go func() {
+		defer writersWG.Done()
+		jupyterStdErr := JupyterStreamWriter{StreamStderr, &receipt}
+		io.Copy(&jupyterStdErr, rErr)
+	}()
+
+	// eval
+	vals, executionErr := k.doEval(code)
```

Comment:
As discussed offline, this should pass code to ipython if there's no kubeflow magic command?

Reply:
We could discuss about it in #8